### PR TITLE
Storage - Improving Perf of StorageBearerTokenChallengeAuthorizationPolicy

### PR DESCRIPTION
**Pull request by ibrandes**
### **Improving Performance of `StorageBearerTokenChallengeAuthorizationPolicy`**

This PR addresses a performance regression introduced in version **12.27.0+** of the Azure Storage SDK for Java, where **token retrieval became thread-local**, causing a **consistent 4–5 second delay** across threads during initial authorization. In contrast, version **12.26.0** exhibited **shared token initialization**, allowing threads to complete their first request within ~250ms of each other.

#### **Root Cause**
`StorageBearerTokenChallengeAuthorizationPolicy.authorizeRequest` always invokes the parent policy via the public `setAuthorizationHeader(...)`, which hard‑codes `checkToForceFetchToken = true`. The base policy’s own `authorizeRequest` path uses `checkToForceFetchToken = false`. That single boolean difference drives a markedly worse concurrency behavior in `AccessTokenCache.retrieveToken(...)` for the storage policy. 

When `checkToForceFetchToken = false`, in `AccessTokenCache.retrieveToken` the [fast path](https://github.com/Azure/azure-sdk-for-java/blob/c92ac941881226258206b3b613527a029eafc6cd/sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCache.java#L156) is never taken. If another thread is proactively refreshing (i.e., wip already set) and the current cached token is still valid, storage callers cannot use that valid token immediately.

Instead, the execution falls into this [else block](https://github.com/Azure/azure-sdk-for-java/blob/c92ac941881226258206b3b613527a029eafc6cd/sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCache.java#L159):
```
else {
    if (checkToForceFetchToken) {
        return Mono.empty();
    }
    ...
}
```
Which will return `Mono.empty()`, triggering the [outer `repeatWhenEmpty` loop](https://github.com/Azure/azure-sdk-for-java/blob/c92ac941881226258206b3b613527a029eafc6cd/sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCache.java#L79) (500 ms delay + resubscribe). Multiple concurrent requests all spin in this empty-repeat cycle waiting for the refresh to finish, despite a perfectly usable (unexpired) token being present. This introduces avoidable latency.

#### **Fix Implemented**
- Refactored `authorizeRequestOnChallenge` and related methods to:
  - Restore **shared token initialization** behavior.
  - Improve parsing of challenge headers (e.g., handling quoted and comma-separated values).
  - Ensure consistent scope and tenant ID extraction from challenge attributes.
  - Introduce `authorizeRequestSync` and `authorizeRequestOnChallengeSync` for synchronous flows.

#### **Performance Validation**
- Post-fix measurements show threads again completing their first request within ~250ms of each other, confirming restored shared initialization.

#### **Code Changes**
- **Files Modified**:
  - `StorageBearerTokenChallengeAuthorizationPolicy.java`: Refactored challenge handling and token request logic.
  - `StorageBearerTokenChallengeAuthorizationPolicyTests.java`: Added/updated unit tests to validate challenge parsing, tenant ID extraction, and multi-tenant scenarios.

### CODE REVIEW

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java [hunk 1]
```diff
@@ -40,134 +42,124 @@ public class StorageBearerTokenChallengeAuthorizationPolicy extends BearerTokenA
...
+            throw LOGGER.logExceptionAsError(new RuntimeException("Invalid authorization URI: tenantId not found"));
+        } catch (URISyntaxException e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException("Invalid authorization URI", e));
+        }
     }
 
-    Map<String, String> extractChallengeAttributes(String header, String authChallengePrefix) {
-        if (!isBearerChallenge(header, authChallengePrefix)) {
+    Map<String, String> extractChallengeAttributes(String header) {
+        if (header == null || !header.regionMatches(true, 0, BEARER_TOKEN_PREFIX, 0, BEARER_TOKEN_PREFIX.length())) {
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java]: There is a newer utility method in `CoreUtils.parseAuthenticateHeader` which could be used to replace this method.

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/main/java/com/azure/core/util/CoreUtils.java#L826

File: sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicyTests.java [hunk 2]
```diff
@@ -3,209 +3,262 @@
...
+
+        // Create a mock pipeline call context
+        HttpPipelineCallContext context = createMockCallContext(request);
+
+        // Test that the policy correctly processes the challenge and returns true
+        boolean result = policy.authorizeRequestOnChallengeSync(context, response);
+
+        assertTrue(result);
+        assertNotNull(request.getHeaders().getValue(HttpHeaderName.AUTHORIZATION));
+        assertTrue(request.getHeaders().getValue(HttpHeaderName.AUTHORIZATION).startsWith("Bearer "));
```
Comment (kyleknap) [inline sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicyTests.java]: Would it be possible to assert aspects such as the tenant id being set or the scopes being updated? Mainly that would add nice end to end coverage from the perspective of the policy class that it is functionally behaving as expected.

File: sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicyTests.java [hunk 3]
```diff
@@ -3,209 +3,262 @@
...
+        Map<String, String> attributes = new HashMap<>();
+        attributes.put("resource_id", "https://storage.azure.com");
 
-        String scope1 = policy.getScopeFromChallenges(challenges1);
-        String authorization1 = policy.getAuthorizationFromChallenges(challenges1);
-        String scope2 = policy.getScopeFromChallenges(challenges2);
-        String authorization2 = policy.getAuthorizationFromChallenges(challenges2);
+        TokenRequestContext result = policy.createTokenRequestContext(attributes);
 
-        assertEquals("https://storage.azure.com", scope1);
```
Comment (kyleknap) [inline sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicyTests.java:158]: Would it make sense to port a test case for multiple different tenants? It's hard to tell the exact scenario this test case is trying to cover because it is only exercising the `extractChallengeAttributes` but curious with the better testing scaffolding that we now have whether we should build a test case that better captures the edge case this test was originally trying to catch.

**Review by kyleknap — COMMENTED** [review]
Looks good. I just had a couple of questions on the tests. We should also make sure to add a changelog entry for this.

File: sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicyTests.java (hunk 3 above)
Comment (ibrandes) [reply sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicyTests.java:158]: I've been trying to add this test, but I think it has shed light onto another token refresh issue - the existing token refresh logic in `AccessTokenCache` cannot take into account `tenantId` when deciding whether to fetch a new token. If the challenges only differ by tenant, only the first token will be fetches and the next two will re-use the cached token. I don't immediately see a simple way to force this refresh without editing `AccessTokenCache` in core, but I will keep looking. For now, I added the test to the PR and added a comment where its failing.

Comment (ibrandes) [reply sdk/storage/azure-storage-common/src/test/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicyTests.java:158]: issue was addressed on core's side and test was added

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java [hunk 4]
```diff
@@ -108,66 +98,60 @@ String extractTenantIdFromUri(String uri) {
...
             return Collections.emptyMap();
         }
 
-        header = header.toLowerCase(Locale.ROOT).replace(authChallengePrefix.toLowerCase(Locale.ROOT), "");
+        // Don't lowercase the entire header as values can be corrupted. Also don't mutate the original header.
+        String remainingHeader = header.substring(BEARER_TOKEN_PREFIX.length());
 
-        String[] attributes = header.split(" ");
+        // Split on commas first; if no commas present fall back to spaces.
+        String[] attributes = remainingHeader.contains(",") ? remainingHeader.split(",") : remainingHeader.split(" ");
```
Comment (kyleknap) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java:135]: What was the reason for splitting on commas in favor of spaces again? I'm slightly concerned that we are changing the precedence here. It might be worth expanding this comment to say 1) why we support commas and 2) why we support both mainly for reference in the future.

Also is it possible that we'd need to handle splitting on both commas and spaces as part of a single auth value? Seems like Python supports [this](https://github.com/Azure/azure-sdk-for-python/blob/d0fc0a64b45b1b1d18a6636734a2c18a468eca02/sdk/eventhub/azure-eventhub-checkpointstoreblob/azure/eventhub/extensions/checkpointstoreblob/_vendor/storage/blob/_shared/authentication.py#L224) but can't really tell if there is an actual use case for it or if just made the code simpler.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java [hunk 5]
```diff
@@ -108,66 +98,60 @@ String extractTenantIdFromUri(String uri) {
...
-            return true;
+        // Create new scopes array to avoid mutating the original
+        String[] scopesToUse = initialScopes;
+        if (!CoreUtils.isNullOrEmpty(resource)) {
+            scopesToUse = new String[] { resource.endsWith(DEFAULT_SCOPE) ? resource : resource + DEFAULT_SCOPE };
         }
 
-        return false;
-    }
+        TokenRequestContext tokenRequestContext = new TokenRequestContext().addScopes(scopesToUse).setCaeEnabled(true);
```
Comment (kyleknap) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java:116]: Is there a particular reason we are calling `setCaeEnabled()`? Mainly noticed it is deviating from how we were building the token request context from before. So, want to make sure we are confident in changing the setting here. It also may be worth adding a comment if the reason is complicated and useful for context.

**Review by kyleknap — COMMENTED** [review]
Looks good! Just had a couple of questions. I like that we've scoped it down much more from before and also added the tests to make sure the policy works as expected end to tend.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java (hunk 5 above)
Comment (ibrandes) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java:116]: `setCaeEnabled()` was something that was missed in the previous version of the policy - it is standard to support this based on the core docs.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java (hunk 4 above)
Comment (ibrandes) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java:135]: previously, java wasn't following the [RFC](https://www.rfc-editor.org/rfc/rfc2616.txt) for the auth header, in which there can be multiple attributes separated by commas or spaces. the pattern here was chosen based on .NET's implementation. there is no case where we'd need to handle splitting on both commas and spaces as part of a single auth value based on the RFC

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java (hunk 5 above)
Comment (kyleknap) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/StorageBearerTokenChallengeAuthorizationPolicy.java:116]: Gotcha. For reference, .NET [set this](https://github.com/Azure/azure-sdk-for-net/blob/ef63c87588495678981a7c1001c019913f1d1158/sdk/storage/Azure.Storage.Common/src/Shared/StorageBearerTokenChallengeAuthorizationPolicy.cs#L64) as well. Also per the [documentation](https://learn.microsoft.com/en-us/java/api/com.azure.core.credential.tokenrequestcontext?view=azure-java-stable#com-azure-core-credential-tokenrequestcontext-setcaeenabled(boolean)), CAE also needs to be enabled on the server end for it to be used. So we are not necessarily forcing this setting downstream if CAE is not desired/supported.
